File: binance4j-websocket/src/main/java/com/binance4j/websocket/serialization/CandlestickEventDeserializer.java
package com.binance4j.websocket.serialization;

import java.io.IOException;

import com.binance4j.websocket.dto.Candle;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;

/** {@link Candle} deserializer */
public class CandlestickEventDeserializer extends JsonDeserializer<Candle> {
	@Override
	public Candle deserialize(JsonParser jp, DeserializationContext ctx) throws IOException {
		JsonNode node = jp.getCodec().readTree(jp);
		JsonNode n = node.get("k");

		return new Candle(n.get("t").asLong(), n.get("o").asText(), n.get("h").asText(), n.get("l").asText(), n.get("c").asText(), n.get("v").asText(),
				n.get("T").asLong(), n.get("q").asText(), n.get("n").asLong(), n.get("V").asText(), n.get("Q").asText(), node.get("e").asText(),
				node.get("E").asLong(), node.get("s").asText(), n.get("i").asText(), n.get("f").asLong(), n.get("L").asLong(), n.get("x").asBoolean());
	}
}

File: connectors/src/main/kotlin/com/binance4j/websocket/serialization/CandlestickEventDeserializer.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.binance4j.websocket.serialization

import com.binance4j.websocket.dto.Candle
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.JsonNode
import java.io.IOException

/** [Candle] deserializer  */
class CandlestickEventDeserializer : JsonDeserializer<Candle>() {
    @Throws(IOException::class)
    override fun deserialize(jp: JsonParser, ctx: DeserializationContext): Candle {
        val node = jp.codec.readTree<JsonNode>(jp)
        val n = node["k"]
        return Candle(
            n["t"].asLong(),
            n["o"].asText(),
            n["h"].asText(),
            n["l"].asText(),
            n["c"].asText(),
            n["v"].asText(),
            n["T"].asLong(),
            n["q"].asText(),
            n["n"].asLong(),
            n["V"].asText(),
            n["Q"].asText(),
            node["e"].asText(),
            node["E"].asLong(),
            node["s"].asText(),
            n["i"].asText(),
            n["f"].asLong(),
            n["L"].asLong(),
            n["x"].asBoolean()
        )
    }
}